janusgraph-cql/src/main/java/org/janusgraph/diskstorage/cql/CQLKeyColumnValueStore.java

```diff
@@ -117,9 +121,37 @@ public class CQLKeyColumnValueStore implements KeyColumnValueStore {
     public static final String KEY_END_BINDING = "keyEnd";
     public static final String LIMIT_BINDING = "maxRows";
 
-    public static final Function<? super Throwable, BackendException> EXCEPTION_MAPPER = cause -> Match(cause).of(
+    public static final Function<? super Throwable, BackendException> EXCEPTION_MAPPER = cause -> {
+        // Unwrap any ExecutionExceptions to get to the real cause:
+        while (cause instanceof ExecutionException) {
+            Throwable t = ((ExecutionException) cause).getCause();
+            if (t == null) {
+                break;
+            } else {
+                cause = t;
+            }
+        }
+        return Match(cause).of(
             Case($(instanceOf(QueryValidationException.class)), PermanentBackendException::new),
+            Case($(CQLKeyColumnValueStore::isKeySizeTooLarge), PermanentBackendException::new),
             Case($(), TemporaryBackendException::new));
+    };
+
+    /**
+     * CQL keys are limited to 64k, and a query that attempts to filter on such
+     * a key with a value that exceeds that length will produce a ServerError
+     * on ScyllaDB, perhaps due to a bug, rather than an InvalidQueryException.
+     * This type of ServerError is a permanent failure, since the same query with
+     * the same value will always produce this error. Properly treating this as a
+     * permanent failure avoids excessive retrying of bad CQL queries and allows
+     * upstream code to properly report the exception.
+     */
+    private static boolean isKeySizeTooLarge(Throwable cause) {
+        AllNodesFailedException failed = ExceptionUtils.throwableOfType(cause, AllNodesFailedException.class);
+        return failed != null && failed.getAllErrors().values().stream().anyMatch(errors ->
+            errors.stream().anyMatch(error -> (error instanceof ServerError)
+                && error.getMessage().startsWith("Key size too large:")));
```

Comment:
For reference, this originates from this bit of code in ScyllaDB: https://github.com/scylladb/scylladb/blob/8d7e35caefecc053ba5495c519cc270632c6db1a/compound.hh#L106